sql/core/src/test/scala/org/apache/spark/sql/DataFrameFunctionsSuite.scala

```diff
@@ -85,6 +86,52 @@ class DataFrameFunctionsSuite extends QueryTest {
     }
   }
 
+  test("concat/concat_ws test") {
+    checkAnswer(
+      testData2.select(concat()).limit(1),
+      Row("")
+    )
+    checkAnswer(
+      testData2.select(
+        concat_ws(lit(","), col("a").cast("String"), col("b").cast("String"))).limit(1),
+      Row("1,1")
+    )
+    checkAnswer(
+      testData2.select(
+        concat_ws(lit("=="), array(col("a").cast("String"), col("b").cast("String")), lit("x"))
+      ).limit(1),
+      Row("1==1==x")
+    )
+    checkAnswer(
+      testData2.select(concat_ws(lit(""), lit("x"), lit("y"))).limit(1),
+      Row("xy")
+    )
+    checkAnswer(
+      testData2.select(concat_ws(lit("=="), array(), lit("x"))).limit(1),
+      Row("==x")
+    )
+    checkAnswer(
+      ctx.sql("""SELECT CONCAT(null, null)"""),
```

Comment:
Nit: `selectExpr` just like the `sql`, will trigger the `FunctionRegistry` eventually, but seems more simple. Anyway it's not a big problem.